Add CI setup that builds and tests Nighthawk using gcc

error msg:

```
/external/envoy/test/mocks/config/BUILD:11:1: C++ compilation of rule '@envoy//test/mocks/config:config_mocks' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF ...

Use --sandbox_debug to see verbose messages from the sandbox
In file included from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/upstream/_virtual_includes/cluster_manager_interface/envoy/upstream/cluster_manager.h:15,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/server/_virtual_includes/factory_context_interface/envoy/server/factory_context.h:26,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/server/_virtual_includes/filter_config_interface/envoy/server/filter_config.h:10,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/config/_virtual_includes/config_provider_manager_interface/envoy/config/config_provider_manager.h:6,
                 from external/envoy/test/mocks/config/mocks.h:3,
                 from external/envoy/test/mocks/config/mocks.cc:1:
bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/config/_virtual_includes/grpc_mux_interface/envoy/config/grpc_mux.h:75:16: error: 'virtual void Envoy::Config::GrpcMux::pause(std::vector<std::__cxx11::basic_string<char> >)' was hidden [-Werror=overloaded-virtual]
   75 |   virtual void pause(const std::vector<std::string> type_urls) PURE;
      |                ^~~~~
In file included from external/com_google_googletest/googlemock/include/gmock/gmock.h:61,
                 from external/envoy/test/test_common/utility.h:35,
                 from external/envoy/test/mocks/config/mocks.h:15,
                 from external/envoy/test/mocks/config/mocks.cc:1:
external/envoy/test/mocks/config/mocks.h:74:21: error:   by 'virtual testing::internal::Function<void(const std::__cxx11::basic_string<char>&)>::Result Envoy::Config::MockGrpcMux::pause(testing::internal::ElemFromListImpl<const std::__cxx11::basic_string<char>&, 0, 0>::type)' [-Werror=overloaded-virtual]
   74 |   MOCK_METHOD(void, pause, (const std::string& type_url));
      |                     ^~~~~
In file included from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/upstream/_virtual_includes/cluster_manager_interface/envoy/upstream/cluster_manager.h:15,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/server/_virtual_includes/factory_context_interface/envoy/server/factory_context.h:26,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/server/_virtual_includes/filter_config_interface/envoy/server/filter_config.h:10,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/config/_virtual_includes/config_provider_manager_interface/envoy/config/config_provider_manager.h:6,
                 from external/envoy/test/mocks/config/mocks.h:3,
                 from external/envoy/test/mocks/config/mocks.cc:1:
bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/config/_virtual_includes/grpc_mux_interface/envoy/config/grpc_mux.h:106:16: error: 'virtual bool Envoy::Config::GrpcMux::paused(std::vector<std::__cxx11::basic_string<char> >) const' was hidden [-Werror=overloaded-virtual]
  106 |   virtual bool paused(const std::vector<std::string> type_urls) const PURE;
      |                ^~~~~~
In file included from external/com_google_googletest/googlemock/include/gmock/gmock.h:61,
                 from external/envoy/test/test_common/utility.h:35,
                 from external/envoy/test/mocks/config/mocks.h:15,
                 from external/envoy/test/mocks/config/mocks.cc:1:
external/envoy/test/mocks/config/mocks.h:76:21: error:   by 'virtual testing::internal::Function<bool(const std::__cxx11::basic_string<char>&)>::Result Envoy::Config::MockGrpcMux::paused(testing::internal::ElemFromListImpl<const std::__cxx11::basic_string<char>&, 0, 0>::type) const' [-Werror=overloaded-virtual]
   76 |   MOCK_METHOD(bool, paused, (const std::string& type_url), (const));
      |                     ^~~~~~
In file included from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/upstream/_virtual_includes/cluster_manager_interface/envoy/upstream/cluster_manager.h:15,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/server/_virtual_includes/factory_context_interface/envoy/server/factory_context.h:26,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/server/_virtual_includes/filter_config_interface/envoy/server/filter_config.h:10,
                 from bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/config/_virtual_includes/config_provider_manager_interface/envoy/config/config_provider_manager.h:6,
                 from external/envoy/test/mocks/config/mocks.h:3,
                 from external/envoy/test/mocks/config/mocks.cc:1:
bazel-out/k8-fastbuild/bin/external/envoy/include/envoy/config/_virtual_includes/grpc_mux_interface/envoy/config/grpc_mux.h:90:16: error: 'virtual void Envoy::Config::GrpcMux::resume(std::vector<std::__cxx11::basic_string<char> >)' was hidden [-Werror=overloaded-virtual]
   90 |   virtual void resume(const std::vector<std::string> type_urls) PURE;
      |                ^~~~~~
In file included from external/com_google_googletest/googlemock/include/gmock/gmock.h:61,
                 from external/envoy/test/test_common/utility.h:35,
                 from external/envoy/test/mocks/config/mocks.h:15,
                 from external/envoy/test/mocks/config/mocks.cc:1:
external/envoy/test/mocks/config/mocks.h:75:21: error:   by 'virtual testing::internal::Function<void(const std::__cxx11::basic_string<char>&)>::Result Envoy::Config::MockGrpcMux::resume(testing::internal::ElemFromListImpl<const std::__cxx11::basic_string<char>&, 0, 0>::type)' [-Werror=overloaded-virtual]
   75 |   MOCK_METHOD(void, resume, (const std::string& type_url));
      |                     ^~~~~~


```

--- a/.bazelrc
+++ b/.bazelrc
@@ -134,23 +134,24 @@ build:coverage --strategy=CoverageReport=sandboxed,local
 build:coverage --experimental_use_llvm_covmap
 build:coverage --collect_code_coverage
 build:coverage --test_tag_filters=-nocoverage
-build:coverage --instrumentation_filter="//source(?!/common/chromium_url|/extensions/quic_listeners/quiche/platform)[/:],//include[/:]"
-coverage:test-coverage --test_arg="--log-path /dev/null"
+build:coverage --instrumentation_filter="//source(?!/extensions/quic_listeners/quiche/platform)[/:],//include[/:]"
 coverage:test-coverage --test_arg="-l trace"
-coverage:fuzz-coverage --config=asan-fuzzer
+coverage:fuzz-coverage --config=plain-fuzzer
 coverage:fuzz-coverage --run_under=@envoy//bazel/coverage:fuzz_coverage_wrapper.sh
 
 # Remote execution: https://docs.bazel.build/versions/master/remote-execution.html
-build:rbe-toolchain --host_platform=@envoy_build_tools//toolchains:rbe_ubuntu_clang_platform
-build:rbe-toolchain --platforms=@envoy_build_tools//toolchains:rbe_ubuntu_clang_platform
 build:rbe-toolchain --action_env=BAZEL_DO_NOT_DETECT_CPP_TOOLCHAIN=1
 
 build:rbe-toolchain-clang --config=rbe-toolchain
+build:rbe-toolchain-clang --platforms=@rbe_ubuntu_clang//config:platform
+build:rbe-toolchain-clang --host_platform=@rbe_ubuntu_clang//config:platform
 build:rbe-toolchain-clang --crosstool_top=@rbe_ubuntu_clang//cc:toolchain
 build:rbe-toolchain-clang --extra_toolchains=@rbe_ubuntu_clang//config:cc-toolchain
 build:rbe-toolchain-clang --action_env=CC=clang --action_env=CXX=clang++ --action_env=PATH=/usr/sbin:/usr/bin:/sbin:/bin:/opt/llvm/bin
 
 build:rbe-toolchain-clang-libc++ --config=rbe-toolchain
+build:rbe-toolchain-clang-libc++ --platforms=@rbe_ubuntu_clang_libcxx//config:platform
+build:rbe-toolchain-clang-libc++ --host_platform=@rbe_ubuntu_clang_libcxx//config:platform
 build:rbe-toolchain-clang-libc++ --crosstool_top=@rbe_ubuntu_clang_libcxx//cc:toolchain
 build:rbe-toolchain-clang-libc++ --extra_toolchains=@rbe_ubuntu_clang_libcxx//config:cc-toolchain
 build:rbe-toolchain-clang-libc++ --action_env=CC=clang --action_env=CXX=clang++ --action_env=PATH=/usr/sbin:/usr/bin:/sbin:/bin:/opt/llvm/bin
@@ -163,6 +164,8 @@ build:rbe-toolchain-msan --linkopt=-Wl,-rpath,/opt/libcxx_msan/lib
 build:rbe-toolchain-msan --config=clang-msan
 
 build:rbe-toolchain-gcc --config=rbe-toolchain
+build:rbe-toolchain-gcc --platforms=@rbe_ubuntu_gcc//config:platform
+build:rbe-toolchain-gcc --host_platform=@rbe_ubuntu_gcc//config:platform
 build:rbe-toolchain-gcc --crosstool_top=@rbe_ubuntu_gcc//cc:toolchain
 build:rbe-toolchain-gcc --extra_toolchains=@rbe_ubuntu_gcc//config:cc-toolchain
 

--- a/bazel/repositories.bzl
+++ b/bazel/repositories.bzl
@@ -1,7 +1,7 @@
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
-ENVOY_COMMIT = "84affaad1173a1e48bbee5b3049c305a555eb3b2"  # July 13th, 2020
-ENVOY_SHA = "df6c0853bbeaf3e36a4db9a5ccbf74a9cff9e436ed52a39aa0d683700a90e4f5"
+ENVOY_COMMIT = "f481e9de7b4754611a7613d8bd9a4a8f146812ec"  # July 14th, 2020
+ENVOY_SHA = "44142dda6fbcebd9b1e62f8c27746cb700db83c871f561de61107d1c360f45ec"
 
 HDR_HISTOGRAM_C_VERSION = "0.9.13"  # Feb 22nd, 2020
 HDR_HISTOGRAM_C_SHA = "2bd4a4631b64f2f8cf968ef49dd03ff3c51b487c3c98a01217ae4cf4a35b8310"

--- a/bazel/repositories.bzl
+++ b/bazel/repositories.bzl
@@ -1,7 +1,7 @@
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
-ENVOY_COMMIT = "f6b86a58b264b46a57d71a9b3b0989b2969df408"  # July 2nd, 2020
-ENVOY_SHA = "5c802266f0cdc5193b6e0247a0f5f20f39f6cc36b688b194e60a853148ba438a"
+ENVOY_COMMIT = "84affaad1173a1e48bbee5b3049c305a555eb3b2"  # July 13th, 2020
+ENVOY_SHA = "df6c0853bbeaf3e36a4db9a5ccbf74a9cff9e436ed52a39aa0d683700a90e4f5"
 
 HDR_HISTOGRAM_C_VERSION = "0.9.13"  # Feb 22nd, 2020
 HDR_HISTOGRAM_C_SHA = "2bd4a4631b64f2f8cf968ef49dd03ff3c51b487c3c98a01217ae4cf4a35b8310"

--- a/bazel/repositories.bzl
+++ b/bazel/repositories.bzl
@@ -1,7 +1,7 @@
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
-ENVOY_COMMIT = "f481e9de7b4754611a7613d8bd9a4a8f146812ec"  # July 14th, 2020
-ENVOY_SHA = "44142dda6fbcebd9b1e62f8c27746cb700db83c871f561de61107d1c360f45ec"
+ENVOY_COMMIT = "7abb0e0bbed4f6b6304403b93762614ad385f80d"  # July 14th, 2020
+ENVOY_SHA = "13fd08f9478e3dee5289581eb8f5f85dfc53fa5ac21555e0e86af536e5a200d8"
 
 HDR_HISTOGRAM_C_VERSION = "0.9.13"  # Feb 22nd, 2020
 HDR_HISTOGRAM_C_SHA = "2bd4a4631b64f2f8cf968ef49dd03ff3c51b487c3c98a01217ae4cf4a35b8310"

--- a/ci/do_ci.sh
+++ b/ci/do_ci.sh
@@ -214,9 +214,7 @@ case "$1" in
     ;;
     test_gcc)
         setup_gcc_toolchain
-        # TODO(#362): change the line below to do_test once the upstream merges
-        # https://github.com/envoyproxy/envoy/pull/10236
-        do_build
+        do_test
         exit 0
     ;;
     clang_tidy)

--- a/source/client/BUILD
+++ b/source/client/BUILD
@@ -46,6 +46,7 @@ envoy_cc_library(
         "//include/nighthawk/common:base_includes",
         "//source/common:request_source_impl_lib",
         "//source/common:nighthawk_common_lib",
+        "@envoy//source/common/common:random_generator_lib_with_external_headers",
         "@envoy//source/common/access_log:access_log_manager_lib_with_external_headers",
         "@envoy//source/common/api:api_lib_with_external_headers",
         "@envoy//source/common/common:cleanup_lib_with_external_headers",

--- a/source/client/benchmark_client_impl.h
+++ b/source/client/benchmark_client_impl.h
@@ -15,6 +15,7 @@
 #include "nighthawk/common/statistic.h"
 
 #include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/random_generator.h"
 #include "external/envoy/source/common/http/http1/conn_pool.h"
 #include "external/envoy/source/common/http/http2/conn_pool.h"
 #include "external/envoy/source/common/runtime/runtime_impl.h"
@@ -129,7 +130,7 @@ private:
   uint32_t max_active_requests_{UINT32_MAX};
   uint32_t max_requests_per_connection_{UINT32_MAX};
   Envoy::Event::TimerPtr timer_;
-  Envoy::Runtime::RandomGeneratorImpl generator_;
+  Envoy::Random::RandomGeneratorImpl generator_;
   uint64_t requests_completed_{};
   uint64_t requests_initiated_{};
   bool measure_latencies_{};

--- a/source/client/process_impl.h
+++ b/source/client/process_impl.h
@@ -16,6 +16,7 @@
 
 #include "external/envoy/source/common/access_log/access_log_manager_impl.h"
 #include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/random_generator.h"
 #include "external/envoy/source/common/event/real_time_system.h"
 #include "external/envoy/source/common/grpc/context_impl.h"
 #include "external/envoy/source/common/http/context_impl.h"
@@ -133,7 +134,7 @@ private:
 
   Envoy::Init::ManagerImpl init_manager_;
   Envoy::LocalInfo::LocalInfoPtr local_info_;
-  Envoy::Runtime::RandomGeneratorImpl generator_;
+  Envoy::Random::RandomGeneratorImpl generator_;
   Envoy::Server::ConfigTrackerImpl config_tracker_;
   Envoy::Secret::SecretManagerImpl secret_manager_;
   Envoy::Http::ContextImpl http_context_;

--- a/source/client/stream_decoder.h
+++ b/source/client/stream_decoder.h
@@ -12,6 +12,7 @@
 #include "nighthawk/common/request_source.h"
 #include "nighthawk/common/statistic.h"
 
+#include "external/envoy/source/common/common/random_generator.h"
 #include "external/envoy/source/common/http/header_map_impl.h"
 #include "external/envoy/source/common/stream_info/stream_info_impl.h"
 #include "external/envoy/source/common/tracing/http_tracer_impl.h"
@@ -42,7 +43,7 @@ public:
                 Statistic& latency_statistic, Statistic& response_header_sizes_statistic,
                 Statistic& response_body_sizes_statistic, HeaderMapPtr request_headers,
                 bool measure_latencies, uint32_t request_body_size,
-                Envoy::Runtime::RandomGenerator& random_generator,
+                Envoy::Random::RandomGenerator& random_generator,
                 Envoy::Tracing::HttpTracerSharedPtr& http_tracer)
       : dispatcher_(dispatcher), time_source_(time_source),
         decoder_completion_callback_(decoder_completion_callback),
@@ -110,7 +111,7 @@ private:
   const uint32_t request_body_size_;
   Envoy::Tracing::EgressConfigImpl config_;
   Envoy::StreamInfo::StreamInfoImpl stream_info_;
-  Envoy::Runtime::RandomGenerator& random_generator_;
+  Envoy::Random::RandomGenerator& random_generator_;
   Envoy::Tracing::HttpTracerSharedPtr& http_tracer_;
   Envoy::Tracing::SpanPtr active_span_;
   Envoy::StreamInfo::UpstreamTiming upstream_timing_;

--- a/source/server/README.md
+++ b/source/server/README.md
@@ -144,9 +144,7 @@ bazel-bin/nighthawk_test_server  [--disable-extensions <string>]
 [--use-fake-symbol-table <bool>]
 [--cpuset-threads]
 [--enable-mutex-tracing]
-[--disable-hot-restart]
-[--max-obj-name-len <uint64_t>]
-[--max-stats <uint64_t>] [--mode
+[--disable-hot-restart] [--mode
 <string>] [--parent-shutdown-time-s
 <uint32_t>] [--drain-strategy <string>]
 [--drain-time-s <uint32_t>]
@@ -192,12 +190,6 @@ Enable mutex contention tracing functionality
 
 --disable-hot-restart
 Disable hot restart functionality
-
---max-obj-name-len <uint64_t>
-Deprecated and unused; please do not specify.
-
---max-stats <uint64_t>
-Deprecated and unused; please do not specify.
 
 --mode <string>
 One of 'serve' (default; validate configs and then serve traffic

--- a/test/BUILD
+++ b/test/BUILD
@@ -238,6 +238,7 @@ envoy_cc_test(
     repository = "@envoy",
     deps = [
         "//source/common:nighthawk_common_lib",
+        "@envoy//source/common/common:random_generator_lib_with_external_headers",
         "@envoy//source/common/stats:isolated_store_lib_with_external_headers",
         "@envoy//test/mocks/local_info:local_info_mocks",
         "@envoy//test/mocks/protobuf:protobuf_mocks",

--- a/test/benchmark_http_client_test.cc
+++ b/test/benchmark_http_client_test.cc
@@ -1,5 +1,6 @@
 #include <vector>
 
+#include "external/envoy/source/common/common/random_generator.h"
 #include "external/envoy/source/common/http/header_map_impl.h"
 #include "external/envoy/source/common/network/utility.h"
 #include "external/envoy/source/common/runtime/runtime_impl.h"
@@ -146,7 +147,7 @@ public:
   Envoy::Stats::IsolatedStoreImpl store_;
   Envoy::Api::ApiPtr api_;
   Envoy::Event::DispatcherPtr dispatcher_;
-  Envoy::Runtime::RandomGeneratorImpl generator_;
+  Envoy::Random::RandomGeneratorImpl generator_;
   NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
   NiceMock<Envoy::Runtime::MockLoader> runtime_;
   std::unique_ptr<Client::BenchmarkClientHttpImpl> client_;

--- a/test/client_worker_test.cc
+++ b/test/client_worker_test.cc
@@ -3,6 +3,7 @@
 
 #include "envoy/upstream/cluster_manager.h"
 
+#include "external/envoy/source/common/common/random_generator.h"
 #include "external/envoy/source/common/runtime/runtime_impl.h"
 #include "external/envoy/source/common/stats/isolated_store_impl.h"
 #include "external/envoy/test/mocks/local_info/mocks.h"
@@ -90,7 +91,7 @@ public:
   MockBenchmarkClient* benchmark_client_;
   MockSequencer* sequencer_;
   MockRequestSource* request_generator_;
-  Envoy::Runtime::RandomGeneratorImpl rand_;
+  Envoy::Random::RandomGeneratorImpl rand_;
   NiceMock<Envoy::Event::MockDispatcher> dispatcher_;
   std::unique_ptr<Envoy::Runtime::ScopedLoaderSingleton> loader_;
   NiceMock<Envoy::LocalInfo::MockLocalInfo> local_info_;

--- a/test/stream_decoder_test.cc
+++ b/test/stream_decoder_test.cc
@@ -1,9 +1,9 @@
 #include <chrono>
 
+#include "external/envoy/source/common/common/random_generator.h"
 #include "external/envoy/source/common/event/dispatcher_impl.h"
 #include "external/envoy/source/common/http/header_map_impl.h"
 #include "external/envoy/source/common/network/utility.h"
-#include "external/envoy/source/common/runtime/runtime_impl.h"
 #include "external/envoy/source/common/stats/isolated_store_impl.h"
 #include "external/envoy/test/mocks/http/mocks.h"
 #include "external/envoy/test/mocks/stream_info/mocks.h"
@@ -49,7 +49,7 @@ public:
   HeaderMapPtr request_headers_;
   uint64_t stream_decoder_completion_callbacks_{0};
   uint64_t pool_failures_{0};
-  Envoy::Runtime::RandomGeneratorImpl random_generator_;
+  Envoy::Random::RandomGeneratorImpl random_generator_;
   Envoy::Tracing::HttpTracerSharedPtr http_tracer_;
   Envoy::Http::ResponseHeaderMapPtr test_header_;
   Envoy::Http::ResponseTrailerMapPtr test_trailer_;

--- a/test/worker_test.cc
+++ b/test/worker_test.cc
@@ -1,5 +1,6 @@
 #include <thread>
 
+#include "external/envoy/source/common/common/random_generator.h"
 #include "external/envoy/source/common/runtime/runtime_impl.h"
 #include "external/envoy/source/common/stats/isolated_store_impl.h"
 #include "external/envoy/test/mocks/local_info/mocks.h"
@@ -36,7 +37,7 @@ public:
   Envoy::Api::ApiPtr api_;
   Envoy::ThreadLocal::MockInstance tls_;
   Envoy::Stats::IsolatedStoreImpl test_store_;
-  Envoy::Runtime::RandomGeneratorImpl rand_;
+  Envoy::Random::RandomGeneratorImpl rand_;
   NiceMock<Envoy::LocalInfo::MockLocalInfo> local_info_;
   NiceMock<Envoy::ProtobufMessage::MockValidationVisitor> validation_visitor_;
 };

